kotlin/workflow-core/src/main/java/com/squareup/workflow/Worker.kt

```diff
@@ -0,0 +1,310 @@
+     * val worker = Worker.create {
+     *   emitOutput(1)
+     *   delay(1000)
+     *   emitOutput(2)
+     *   delay(1000)
+     *   emitOutput(3)
+     * }
+     * ```
+     */
+    inline fun <reified T> create(
+      key: String = "",
+      noinline block: suspend Emitter<T>.() -> Unit
+    ): Worker<T> = TypedWorker(T::class, key, block)
+
+    /**
+     * Creates a [Worker] from a function that returns a single value.
+     *
+     * The returned [Worker] will equate to any other workers created with any of the [Worker]
+     * builder functions that have the same output type.
+     */
+    inline fun <reified T> from(
+      key: String = "",
+        // This could be crossinline, but there's a coroutines bug that will cause the coroutine
+        // to immediately resume on suspension inside block when it is crossinline.
+        // TODO link to KT ticket.
+      noinline block: suspend () -> T
+    ): Worker<T> = create(key) {
+      emitOutput(block())
+    }
+
+    /**
+     * Creates a [Worker] from a function that returns a single value.
+     * The worker will emit the value **if and only if the value is not null**, then finish.
+     *
+     * The returned [Worker] will equate to any other workers created with any of the [Worker]
+     * builder functions that have the same output type.
+     */
+    inline fun <reified T : Any> fromNullable(
+      key: String = "",
+        // This could be crossinline, but there's a coroutines bug that will cause the coroutine
+        // to immediately resume on suspension inside block when it is crossinline.
+        // TODO link to KT ticket.
+      noinline block: suspend () -> T?
+    ): Worker<T> = create(key) {
+      block()?.let { emitOutput(it) }
+    }
+
+    /**
+     * Creates a [Worker] from a function that returns a [ReceiveChannel].
+     * The worker will emit everything from the returned channel, and finish when the channel is
+     * closed.
+     *
+     * If the [Worker] is torn down before finishing, both the scope and the channel will be
+     * cancelled.
+     *
+     * The returned [Worker] will equate to any other workers created with any of the [Worker]
+     * builder functions that have the same output type.
+     */
+    inline fun <reified T> fromChannel(
+      key: String = "",
+      crossinline block: CoroutineScope.() -> ReceiveChannel<T>
+    ): Worker<T> = create(key) {
+      coroutineScope {
+        val channel = block()
+        // Using consumeEach ensures that the channel is closed if this coroutine is cancelled.
+        // This is important if block doesn't use the passed CoroutineScope to scope the returned
+        // channel.
+        @Suppress("EXPERIMENTAL_API_USAGE")
+        channel.consumeEach {
+          emitOutput(it)
+        }
+      }
+    }
+  }
+}
+
+/**
+ * Returns a [Worker] that will await this [Deferred] and then emit it.
+ *
+ * Note that [Deferred] is a "hot" future type – calling a function that _returns_ a [Deferred]
```

Comment:
Is this caveat a good reason to get rid of this extension entirely? Wondering if it's going to be easier to misuse than  not.

Reply:
Seems similar to the concern you raise in #337